src/main/kotlin/com/github/avrokotlin/avro4k/AnnotationExtractor.kt

```diff
@@ -1,28 +1,65 @@
+    fun jsonProps(): List<Pair<String, String>> = annotations.filterIsInstance<AvroJsonProp>().map { it.key to it.jsonValue }
+    fun default(): String? = annotations.filterIsInstance<AvroDefault>().firstOrNull()?.value
+    fun enumDefault(): String? = annotations.filterIsInstance<AvroEnumDefault>().firstOrNull()?.value
+}
+
+@ExperimentalSerializationApi
+internal fun SerialDescriptor.getAvroFullName(namingStrategy: NamingStrategy): String {
+    val namespace = annotations.firstOrNull { it is AvroNamespace }?.let { (it as AvroNamespace).value }
+    val name = annotations.firstOrNull { it is AvroName }?.let { namingStrategy.to((it as AvroName).value) }
+    return if (namespace != null) {
+        if (name != null) {
+            "$namespace$name"
+        } else {
+            val serialOnlyName = namingStrategy.to(serialName.substringAfterLast('.').let { if (isNullable) it.removeSuffix('?') else it })
+            "$namespace$serialOnlyName"
+        }
+    } else {
+        if (name != null) {
+            "$name"
+        } else {
+            val dotSeparatorIndex = serialName.lastIndexOf('.')
+            "${serialName.substring(0, dotSeparatorIndex)}.${namingStrategy.to(serialName.substring(dotSeparatorIndex + 1).let { if (isNullable) it.removeSuffix('?') else it })}"
+        }
+    }
+}
 
-      operator fun invoke(descriptor: SerialDescriptor, index: Int): AnnotationExtractor =
-         AnnotationExtractor(descriptor.getElementAnnotations(index))
-   }
+@ExperimentalSerializationApi
+fun String.removeSuffix(char: Char): String =
+        if (endsWith(char)) substring(0, length - 1) else this
 
-   fun fixed(): Int? = annotations.filterIsInstance<AvroFixed>().firstOrNull()?.size
-   fun scalePrecision(): Pair<Int, Int>? = annotations.filterIsInstance<ScalePrecision>().firstOrNull()?.let { it.scale to it.precision }
-   fun namespace(): String? = annotations.filterIsInstance<AvroNamespace>().firstOrNull()?.value
-   fun name(): String? = annotations.filterIsInstance<AvroName>().firstOrNull()?.value
-   fun valueType(): Boolean = annotations.filterIsInstance<AvroInline>().isNotEmpty()
-   fun doc(): String? = annotations.filterIsInstance<AvroDoc>().firstOrNull()?.value
-   fun aliases(): List<String> = (annotations.firstNotNullOfOrNull { it as? AvroAlias }?.value ?: emptyArray()).asList() + (annotations.firstNotNullOfOrNull {it as? AvroAliases}?.value ?: emptyArray())
-   fun props(): List<Pair<String, String>> = annotations.filterIsInstance<AvroProp>().map { it.key to it.value }
-   fun jsonProps(): List<Pair<String, String>> = annotations.filterIsInstance<AvroJsonProp>().map { it.key to it.jsonValue }
-   fun default(): String? = annotations.filterIsInstance<AvroDefault>().firstOrNull()?.value
-   fun enumDefault(): String? = annotations.filterIsInstance<AvroEnumDefault>().firstOrNull()?.value
+@ExperimentalSerializationApi
+internal fun SerialDescriptor.getElementAvroName(namingStrategy: NamingStrategy, index: Int): String {
```

Comment:
I prefer `descriptor.getElementAvroName(namingStrategy, index)` than old java helper class fashion `AnnotationExtractor(descriptor.getElementAnnotations(index)).avroName/namespace...`.

I would do the same for all other methods here to have great extensions